blogposts/2021/optimizing-a-code-intelligence-commit-graph.md

```diff
@@ -0,0 +1,699 @@
+
+#### Example
+
+The following hypothetical commit graph contains a number of feature branches that are eventually merged back into mainline, unlike our previous example where all commits had at most one parent.
+
+<div class="no-shadow">
+  <img src="https://sourcegraphstatic.com/blog/commit-graph-optimizations/graph2.png" alt="sample commit graph">
+</div>
+
+Running the query above from the commit `703e33` produces the following CTE results over the first four iterations.
+
+<table>
+<tr><th>iteration</th><th>commit</th><th>parent_commit</th><th>has_lsif_data</th><th>distance</th><th>direction</th></tr>
+<tr><td>1</td><td><code>703e33</code></td><td><code>6307e6</code></td><td>false</td><td>0</td><td>A</td></tr>
+<tr><td>1</td><td><code>703e33</code></td><td><code>6307e6</code></td><td>false</td><td>0</td><td>D</td></tr>
+<tr><td>1</td><td><code>703e33</code></td><td><code>5a24e4</code></td><td>false</td><td>0</td><td>A</td></tr>
+<tr><td>1</td><td><code>703e33</code></td><td><code>5a24e4</code></td><td>false</td><td>0</td><td>D</td></tr>
+<tr><td>2</td><td><code>5a24e4</code></td><td><code>3d2f27</code></td><td>false</td><td>1</td><td>A</td></tr>
+<tr><td>2</td><td><code>6307e6</code></td><td><code>4a848f</code></td><td>false</td><td>1</td><td>A</td></tr>
+<tr class="workingtable-highlight"><td>3</td><td><code>3d2f27</code></td><td><code>09210f</code></td><td>false</td><td>2</td><td>A</td></tr>
+<tr class="workingtable-highlight"><td>3</td><td><code>3d2f27</code></td><td><code>2190d3</code></td><td>false</td><td>2</td><td>A</td></tr>
+<tr><td>3</td><td><code>4a848f</code></td><td><code>3d2f27</code></td><td>false</td><td>2</td><td>A</td></tr>
+<tr><td>4</td><td><code>09210f</code></td><td></td><td>false</td><td>3</td><td>A</td></tr>
+<tr><td>4</td><td><code>2190d3</code></td><td><code>1f64f9</code></td><td>false</td><td>3</td><td>A</td></tr>
+<tr class="workingtable-highlight"><td>4</td><td><code>3d2f27</code></td><td><code>09210f</code></td><td>false</td><td>3</td><td>A</td></tr>
+<tr class="workingtable-highlight"><td>4</td><td><code>3d2f27</code></td><td><code>2190d3</code></td><td>false</td><td>3</td><td>A</td></tr>
+</table>
+
+Notice that there are two ways to get from commit `7033ee` to commit `3d2f27`, the therefore the entries for `3d2f27` are duplicated in the CTE results (and therefore the working table). Also notice that the number of _new_ rows per iteration is growing as the iteration count rises in such graphs. **For some configuration of input, this query is [quadratic](https://accidentallyquadratic.tumblr.com/) instead of linear.**
+
+What we _wanted_ to happen was for the "duplicate rows" not to be inserted into the working table at all. Unfortunately, each record in this example distinct due to the differing path lengths - a detail we glossed over when designing this query in the first place. Another classic case of the computer doing what you _told_ it to do instead of what you _wanted_ it to do. 🙄
```

Comment:
```suggestion
What we _wanted_ to happen was for the "duplicate rows" not to be inserted into the working table at all. Unfortunately, each record in this example is distinct due to the differing path lengths - a detail we glossed over when designing this query in the first place. Another classic case of the computer doing what you _told_ it to do instead of what you _wanted_ it to do. 🙄
```